kong/tools/database_cache.lua

```diff
@@ -49,28 +52,62 @@ end
 
 local DATA = {}
 
-function _M.set(key, value)
+function _M.set(key, value, exptime)
+  exptime = exptime or 0
+  
+  if exptime ~= 0 then
+    value = {
+      value = value,
+      [TTL_EXPIRE_KEY] = gettime() + exptime,
+    }
+  end
+
   DATA[key] = value
 
   -- Save into Shared Dictionary
-  local _, err = _M.sh_set(key, cjson.encode(value))
+  local _, err = _M.sh_set(key, cjson.encode(value), exptime)
   if err then return nil, err end
 
   return true
 end
 
 function _M.get(key)
-  local value = DATA[key]
+  local now = gettime()
   
-  -- Get from Shared Dictionary
-  if value == nil then
-    value = _M.sh_get(key)
-    if value ~= nil then 
-      value = cjson.decode(value) 
+  -- check local memory, and verify ttl
+  local value = DATA[key]
+  if value ~= nil then
+    if type(value) ~= "table" or not value[TTL_EXPIRE_KEY] then
+      -- found non-ttl value, just return it
+      return value
+    elseif value[TTL_EXPIRE_KEY] >= now then
+      -- found ttl-based value, within ttl
+      return value.value
+    else
+      -- value with expired ttl, delete it
+      DATA[key] = nil
+      -- value = nil  -- commented out because it triggers a linter error
     end
   end
-
-  return value
+  
+  -- nothing found yet, get it from Shared Dictionary
+  value = _M.sh_get(key)
+  if not value then
+    -- nothing found
+    return nil
+  else
+    value = cjson.decode(value)
+    DATA[key] = value  -- store in memory, so we don't need to deserialize next time
+
+    if type(value) ~= "table" or not value[TTL_EXPIRE_KEY] then
```

Comment:
imo, mixing tables and plain values leads to really confusing code, and forces us to deal with such corner-cases. Ex: checking for return type, checking for `expire`, returning `value.value` or just `value`... All of that not only is confusing to the reader and can easily result in a "disaster waiting to happen", but it also prevents proper branch optimization, and just plain old optimization, for that matter (using `cjson.encode()` when the stored values are plain scalars...).